### PR TITLE
machinst x64: implement calls and int cmp/store/loads;

### PR DESCRIPTION
**Pull request by bnjbvr**
This makes it possible to run a simple recursive fibonacci function in
wasmtime, with the new x64 backend. Plenty of todos implemented as assertions that may and will trigger on larger programs, plenty of paths are still untested, still a lot of missing opcodes lowering and code gen, plenty of optimizations, etc etc.

### CODE REVIEW

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 1]
```diff
@@ -1,37 +1,59 @@
...
+    /// Argument locations (regs or stack slots). Stack offsets are relative to
+    /// SP on entry to function.
+    args: Vec<ABIArg>,
+    /// Return-value locations. Stack offsets are relative to the return-area
+    /// pointer.
+    rets: Vec<ABIArg>,
+    /// Space on stack used to store arguments.
+    stack_arg_space: i64,
+    /// Space on stack used to store return values.
+    stack_ret_space: i64,
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/abi.rs:38]: Does space mean size or location?

Comment (cfallin) [reply cranelift/codegen/src/isa/x64/abi.rs:38]: Size, in this case (the `aarch64` impl is the same (and as an aside, we hope to factor out the commonalities here soon)).

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 2]
```diff
@@ -520,3 +551,482 @@ impl ABIBody for X64ABIBody {
...
+        B8 | I8 => 8,
+        B16 | I16 => 16,
+        B32 | I32 | F32 => 32,
+        B64 | I64 | F64 => 64,
+        B128 | I128 => 128,
+        IFLAGS | FFLAGS => 32,
+        B8X8 | I8X8 | B16X4 | I16X4 | B32X2 | I32X2 => 64,
+        B8X16 | I8X16 | B16X8 | I16X8 | B32X4 | I32X4 | B64X2 | I64X2 => 128,
+        _ => panic!("ty_bits() on unknown type: {:?}", ty),
+    }
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/abi.rs:569]:
```suggestion
    ty.bits()
```

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 1 above)
Comment (bjorn3) [reply cranelift/codegen/src/isa/x64/abi.rs:38]: Why is it an `i64` then? A negative size doesn't make sense.

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 3]
```diff
@@ -37,61 +35,71 @@ use regs::{create_reg_universe_systemv, show_ireg_sized};
...
-    MovSX_M_R {
-        extMode: ExtMode,
-        addr: Addr,
+    /// Loads the memory address of addr into dst.
+    LoadEffectiveAddress {
+        addr: SyntheticAmode,
+        dst: Writable<Reg>,
+    },
+
+    /// Sign-extended loads: movs (bl bq wl wq lq) addr reg.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:88]: pico-nit (smaller even than a nano-nit): These are now sign-extended loads *and* moves.

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 4]
```diff
@@ -100,70 +108,92 @@ pub(crate) enum Inst {
...
     /// pushq (reg addr imm)
     Push64 { src: RegMemImm },
 
     /// popq reg
     Pop64 { dst: Writable<Reg> },
 
-    /// call simm32
+    // =====================================
+    // Floating-point operations.
+    /// Float arithmetic/bit-twiddling: (add sub and or xor mul adc? sbb?) (32 64) (reg addr imm) reg
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:131]: atto-nit: (reg addr), but no imm

File: cranelift/codegen/src/isa/x64/inst/mod.rs [hunk 5]
```diff
@@ -100,70 +108,92 @@ pub(crate) enum Inst {
         dst: Writable<Reg>,
     },
 
-    /// cmp (b w l q) (reg addr imm) reg
+    /// Integer comparisons: cmp (b w l q) (reg addr imm) reg.
```
Comment (julian-seward1) [inline cranelift/codegen/src/isa/x64/inst/mod.rs:111]: Random comment: if at some point you add `tst` as an instruction, it has the same "shape" as `cmp` and so would fit naturally here.  (cmp = subtract operands, set flags, throw away result; tst = AND operands; set flags; throw away result).

**Comment by julian-seward1** [comment]
This looks great to me; thanks for having the patience to iterate it.  Land, I say!

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 2 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs]: Thanks! Even updated the callers so they use ty.bits() instead.

File: cranelift/codegen/src/isa/x64/abi.rs (hunk 1 above)
Comment (bnjbvr) [reply cranelift/codegen/src/isa/x64/abi.rs:38]: Good question; I'll take care of it during the Grand Refactoring.

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 6]
```diff
@@ -132,97 +154,50 @@ fn get_callee_saves(regs: Vec<Writable<RealReg>>) -> Vec<Writable<RealReg>> {
...
-
-                _ => {
-                    unimplemented!("non normal argument purpose");
-                }
-            }
-        }
+        let call_conv = f.signature.call_conv;
+        debug_assert!(
+            call_conv == isa::CallConv::SystemV || call_conv.extends_baldrdash(),
+            "unsupported or unimplemented calling convetion {}",
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/abi.rs:163]: *convention

File: cranelift/codegen/src/isa/x64/abi.rs [hunk 7]
```diff
@@ -524,3 +551,465 @@ impl ABIBody for X64ABIBody {
...
+        let (next_reg, candidate) = if intreg {
+            let candidate = match args_or_rets {
+                ArgsOrRets::Args => get_intreg_for_arg_systemv(next_gpr),
+                ArgsOrRets::Rets => get_intreg_for_retval_systemv(next_gpr),
+            };
+            (&mut next_gpr, candidate)
+        } else {
+            let candidate = match args_or_rets {
+                ArgsOrRets::Args => get_fltreg_for_arg_systemv(next_gpr),
+                ArgsOrRets::Rets => get_fltreg_for_retval_systemv(next_gpr),
```
Comment (bjorn3) [inline cranelift/codegen/src/isa/x64/abi.rs:702]: This should have used `next_vreg` instead of `next_gpr`.

**Pull request merged**